src/main/java/org/opentripplanner/routing/stoptimes/AlternativeLegs.java

```diff
@@ -0,0 +1,221 @@
+            RoutingService routingService,
+            TimetableSnapshot timetableSnapshot,
+            T2<TripPattern, P2<Integer>> patternWithBoardAlightPositions,
+            Calendar departureTime,
+            ServiceDate originalDate,
+            boolean searchBackward
+    ) {
+        var pattern = patternWithBoardAlightPositions.first;
+        var boardingPosition = patternWithBoardAlightPositions.second.first;
+        var alightingPosition = patternWithBoardAlightPositions.second.second;
+
+        // TODO: What should we have here
+        ZoneId timeZone = routingService.getTimeZone().toZoneId();
+
+        Comparator<TripTimeOnDate> comparator = Comparator.comparing((TripTimeOnDate tts) -> tts.getServiceDayMidnight() + tts.getRealtimeDeparture());
+
+        if (searchBackward) {
+            comparator = comparator.reversed();
+        }
+
+        Queue<TripTimeOnDate> pq = new PriorityQueue<>(comparator);
+
+        long startTime = departureTime.getTimeInMillis() / 1000;
+
+        // Loop through all possible days
+        ServiceDate[] serviceDates = {originalDate.previous(), originalDate, originalDate.next()};
+
+        for (ServiceDate serviceDate : serviceDates) {
+            ServiceDay sd = new ServiceDay(
+                    routingService.getServiceCodes(),
+                    serviceDate,
+                    routingService.getCalendarService(),
+                    pattern.getRoute().getAgency().getId()
+            );
+
+            Timetable timetable;
+            if (timetableSnapshot != null) {
+                timetable = timetableSnapshot.resolve(pattern, serviceDate);
+            }
+            else {
+                timetable = pattern.getScheduledTimetable();
+            }
+
+            int secondsSinceMidnight = sd.secondsSinceMidnight(startTime);
+            for (TripTimes tripTimes : timetable.getTripTimes()) {
+                if (!sd.serviceRunning(tripTimes.getServiceCode())) {continue;}
+                if (skipByTripCancellation(tripTimes, false)) {continue;}
+
+                boolean departureTimeInRange = searchBackward
+                        ? tripTimes.getDepartureTime(boardingPosition) <= secondsSinceMidnight
+                        : tripTimes.getDepartureTime(boardingPosition) >= secondsSinceMidnight;
+
+                if (departureTimeInRange) {
+                    pq.add(new TripTimeOnDate(tripTimes, boardingPosition, pattern, sd));
+                }
+            }
+
+        }
+
+        List<ScheduledTransitLeg> res = new ArrayList<>();
+
+        while(!pq.isEmpty()){
+            TripTimeOnDate tripTimeOnDate = pq.poll();
+            res.add(mapToLeg(timeZone, pattern, boardingPosition, alightingPosition, tripTimeOnDate));
+        }
+
+        return res.stream();
+    }
+
+    @Nonnull
+    private static ScheduledTransitLeg mapToLeg(
+            ZoneId timeZone,
+            TripPattern pattern,
+            Integer boardingPosition,
```

Comment:
Can these be null or why aren't they `int`?

Reply:
fixed